pkg/blobinfocache/internal/prioritize/prioritize.go

```diff
@@ -77,34 +82,58 @@ func (css *candidateSortState) Swap(i, j int) {
 	css.cs[i], css.cs[j] = css.cs[j], css.cs[i]
 }
 
-// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with a parameter for the
-// number of entries to limit, only to make testing simpler.
-func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidates int) []blobinfocache.BICReplacementCandidate2 {
+// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with parameters for the
+// number of entries to limit for known and unknown location separately, only to make testing simpler.
+func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidates int, unknownLocationLimit int) []blobinfocache.BICReplacementCandidate2 {
+	// split unknown candidates and known candidates
+	// and limit them seperately.
+	var knownLocationCandidates []CandidateWithTime
+	var unknownLocationCandidates []CandidateWithTime
+
+	for _, candidate := range cs {
+		if candidate.Candidate.UnknownLocation {
+			unknownLocationCandidates = append(unknownLocationCandidates, candidate)
+		} else {
+			knownLocationCandidates = append(knownLocationCandidates, candidate)
+		}
+	}
+
 	// We don't need to use sort.Stable() because nanosecond timestamps are (presumably?) unique, so no two elements should
 	// compare equal.
 	sort.Sort(&candidateSortState{
```

Comment:
`unknownLocationCandidates` should still be sorted, primarily so that we can deprioritize `uncompressedDigest`, and also to make the output deterministic and thus testable.

Reply:
done.